### ISSUE TITLE
Add support for upcoming packed arrays tags

### ISSUE THREAD

**Issue by neothemachine** [FasterXML/jackson-dataformat-cbor#12]
There is a draft RFC to be published soon which adds support for packed/typed arrays in CBOR, see the [current version](https://tools.ietf.org/html/draft-jroatch-cbor-tags-02). As one of the original CBOR authors is also an author of this proposal I'm sure it will gain widespread adoption in encoders and decoders, especially because it is such a useful feature and will help greatly when big packed arrays need to be efficiently encoded and also decoded both in terms of space usage and speed.

I thought it would be good putting it on the radar for implementation. I think it will be rather straightforward to implement. I currently don't have time for it myself in the next 3-4 weeks but I certainly will help where I can, reviewing code etc.

**Comment by cowtowncoder** [comment]
Thank you for filing the request. Sounds like something that would be good support.

Handling of packed arrays is not completely trivial, as it deviates from logical model of JSON. I am actually slightly surprised that it would be added, since while I understand benefits, I am not sure those are significant enough to warrant the change.
The main problem really being that the concept of uniform typing in Arrays does not exist in JSON, and as a result, this becomes model level incompatiblity between JSON and CBOR.

Still, it is doable, and `jackson-dataformat-protobuf` (for example) implements functionality already.
Bit more state will be needed for both encoder and decoder, as well as error checking on encoder.

**Comment by neothemachine** [comment]
I work on a project which uses CBOR for encoding scientific data for the web and currently the lack of packed array support is a somehow big problem for us. Not just because data gets potentially bigger but also because it takes a lot more time to read big non-typed arrays into corresponding Javascript arrays. With packed arrays, if endianness matches, you can simply create a typed array view of the range of bytes of the CBOR document, which is pretty much a no-op. Sometimes gzipping brings down a 50MiB CBOR to a few MiB (depending on the data of course) but the parsing then still takes seconds. So for us the benefits seem quite significant and having packed array support keeps us from using alternatives like Google's protocol buffers format.

**Comment by cowtowncoder** [comment]
@neothemachine I understand the potential technical benefits. It's just that to me the cost of lost interoperability with JSON is still a big downside: full 100% accurate round-tripping will not be possible.
Although I guess logical contents can still be transformed, only packed/unpacked information may be lost.

But As I said, this does not mean that it couldn't be supported; if it's in the spec, we'll make sure it is supported so that interoperability between CBOR codecs remains.

One practical question: I wonder how do javascript encoders make use of this feature? I guess code will need to traverse arrays to ensure uniformess, since there are no strongly-typed (wrt element type) arrays in Javascript. I also assume that packing is probably only used for primitive types and Strings.

For Jackson it will be easier to add support for decoding: for encoding same question applies; if and how to detect arrays that may be packed. At streaming level there is no type information to use (unlike at `jackson-databind` level).
Actually, come to think of this, it may not be possible to efficiently support encoding at all, without buffering all contents of the array. So the first phase will be to support decoding, and only then (if then) consider encoding.
Difference from protobuf is that due to mandatory schema, which is carried along to streaming encoder (`JsonGenerator`), information on packing is actually available.

**Comment by neothemachine** [comment]
> One practical question: I wonder how do javascript encoders make use 
> of this feature? I guess code will need to traverse arrays to ensure 
> uniformess, since there are no strongly-typed (wrt element type) 
> arrays in Javascript. I also assume that packing is probably only used 
> for primitive types and Strings.
Well JavaScript actually has typed arrays 
(https://developer.mozilla.org/de/docs/Web/JavaScript/Typed_arrays) and 
only when those are used I would as an encoder use the typed array 
encoding for CBOR. I wouldn't scan regular arrays as this is too 
resource-intensive. For decoding you may argue that you really only want 
JSON compatible JS objects (without typed arrays), so there may be an 
option for decoding where you can switch to the typed-array mode which 
means that CBOR typed arrays are directly mapped to efficient JS typed 
arrays.

> Actually, come to think of this, it may not be possible to efficiently 
> support encoding at all, without buffering all contents of the array. 
> So the first phase will be to support decoding, and only then (if 
> then) consider encoding.
>
Similar to JS I would only encode typed arrays if the Java type is 
actually a primitive array like int[]. List<Integer> is not suitable as 
it can have nulls. If there is really need for a streaming solution for 
typed arrays then that can always be built in later I guess with some 
lower level API where the data provider has to promise to only deliver 
values of the same primitive data type, otherwise it bombs out with an 
exception which is fine as it is clearly a programming error. Does that 
make sense?

**Comment by cowtowncoder** [comment]
Interesting (wrt native typed arrays). It's been a while since I programmed in Javascript.
I did not know that type system / representation has evolved since early days, good to know.

As to streaming: I do not mean that there is need for streaming as much as the fact that CBOR encoder only implements encoding and decoding via `JsonParser` and `JsonGenerator` abstractions -- it does not know anything about higher level constructs, if any. So all it sees for encoding is `START_ARRAY`, followed by zero or more value tokens, and `END_ARRAY`. This is sufficient to encode arbitrary JSON, which was the original use case.

So the challenge would be to either detect sequence of things (all values are `VALUE_NUMBER_INT` tokens, for example) dynamically; or, try to pipe bit of type hint from databinding level.
Latter could be useful for other use cases as well, so perhaps that would be something to consider -- it could help with XML (for example), as well as some databinding use cases.
